### ISSUE TITLE
Expand on second level causing error if more than one properties are included in expand on the first level

### ISSUE THREAD

**Issue by mmichtch** [OData/WebApi#2017]
Expand of the property(properties) on the sub-object is failing if more than one object is included in the expand directive on the first level. 

### Assemblies affected

Microsoft.AspNetCore.OData 7.3.0
Microsoft.EntityFrameworkCore.SqlServer  3.1.0

### Reproduce steps

Send HTTP get request to the following URL: 
https://localhost:5001/odata/Books?$expand=Authors,CookBook($expand=Recipes)

### Expected result

Service should render object with two sub-objects and expand collection property in the respected sub-object.

### Actual result

Application failing with exception:
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index') 

### Additional detail
following requests are working as expected:
https://localhost:5001/odata/Books?$expand=Authors,CookBook
https://localhost:5001/odata/Books?$expand=CookBook($expand=Recipes)

you can find source code for the sample project in:
https://github.com/mmichtch/ODataSample

**Comment by mmichtch** [comment]
Lazy load solves this problem as well as #2016. There is mentioning the discussion about disadvantages of the Lazy load,so far I could not fine any ill effects. SQL seems to be of the same quality and performance does not show any symptoms of degradation.
still, it looks like a bug, without lazy load

**Comment by mmichtch** [comment]
**Correction, Lazy Load did not solve issue completely.** 
I have added another collection property on the second level object and now following requests **are not working**:
https://localhost:5001/odata/Books?$expand=Authors,CookBook($expand=Recipes)
https://localhost:5001/odata/Books?$expand=CookBook($expand=Recipes,BlogEntries)

At the same time following requests are working as expected:
https://localhost:5001/odata/Books?$expand=CookBook,Genres
https://localhost:5001/odata/Books?$expand=CookBook($expand=Recipes)
https://localhost:5001/odata/Books?$expand=CookBook($expand=BlogEntries)

**Bottom line with or without Lazy Load there is a bug related in implementation of the "$expand"  directive.**

**Comment by xuzhg** [comment]
It seems it's same as #2016.

**Comment by mmichtch** [comment]
Yes, it does look very similar to #2016, but this one concentrates on expands on the sub-objects. It is difficult to see from outside what specific parts of the framework are failing, and if internally it is same mechanism. But if those two issues will be merged, please make sure that scenario described here is also tested.

**Comment by AmraniO** [comment]
I'm facing the same issue, can't migrate my project to .NET Core 3.1.

**Comment by asyncoder** [comment]
FYI

https://github.com/dotnet/efcore/issues/19616#issuecomment-586447790

**Comment by asyncoder** [comment]
@xuzhg do you have any updates regarding this? This is a blocker for many people including me willing to upgrade to .NET Core 3.x

**Comment by keatkeat87** [comment]
same problem here, any workaround ?

**Comment by mmichtch** [comment]
FYI, just updated Microsoft.AspNetCore.OData to 7.4.0 - still the same error.
are there any updates on this issue?

**Comment by mmichtch** [comment]
Just updated test application to with preview versions of the components:

- Microsoft.AspNetCore.OData Version="7.4.0"
- Microsoft.EntityFrameworkCore.InMemory Version="5.0.0-preview.3.20181.2"
- Microsoft.EntityFrameworkCore.SqlServer  Version="5.0.0-preview.3.20181.2"
- Microsoft.EntityFrameworkCore.Tools  Version="5.0.0-preview.3.20181.2


error resolved

**Comment by xuzhg** [comment]
Close it since it got fixed after update the EF Core library.
